src/nvim/eval/encode.c

```diff
@@ -315,6 +315,56 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
       ga_append(gap, ')'); \
     } while (0)
 
+#define TYPVAL_ENCODE_CONV_PARTIAL(pt) \
+    do { \
+        int i; \
+        ga_concat(gap, "function("); \
+        if (&pt->pt_name != NULL) { \
+          unsigned len; \
+          char_u *p; \
+          len = 3; \
+          len += (unsigned)STRLEN(pt->pt_name); \
+          for (p = pt->pt_name; *p != NUL; mb_ptr_adv(p)) { \
+            if (*p == '\'') { \
+              len++; \
+            } \
+          } \
+          char_u *r, *s; \
+          s = r = xmalloc(len); \
+          if (r != NULL) { \
+            *r++ = '\''; \
+            for (p = pt->pt_name; *p != NUL; ) { \
+              if (*p == '\'') { \
+                *r++ = '\''; \
+              } \
+            MB_COPY_CHAR(p, r); \
+            } \
+            *r++ = '\''; \
+            *r++ = NUL; \
+          } \
+          ga_concat(gap, s); \
+          xfree(s); \
+        } \
+        if (pt->pt_argc > 0) { \
+          ga_concat(gap, ", ["); \
+          for (i = 0; i < pt->pt_argc; i++) { \
+            if (i > 0) { \
+              ga_concat(gap, ", "); \
+            } \
+            ga_concat(gap, encode_tv2string(&pt->pt_argv[i], NULL)); \
```

Comment:
These calls to `encode_tv2string()` are at least part of the leak. `encode_tv2string()` allocates, then `ga_concat()` copies the result without freeing. Try this instead:

```c
    char *tofree = encode_tv2string(&pt->pt_argv[i], NULL);                                                                  
    ga_concat(gap, tofree);
    xfree(tofree);
```

Same for the other calls.